tools/cosmovisor/process_test.go

```diff
@@ -168,6 +169,103 @@ func (s *processTestSuite) TestLaunchProcessWithDownloads() {
 	require.Equal(cfg.UpgradeBin("chain3"), currentBin)
 }
 
+// TestCustomPreupgrade will try running the script a few times and watch upgrades work properly
+// and args are passed through
+func (s *processTestSuite) TestLaunchProcessWithDownloadsAndMissingPreupgrade() {
+	// test case upgrade path (binaries from testdata/download directory):
+	// genesis -> chain2-zip_bin
+	// chain2-zip_bin -> ref_to_chain3-zip_dir.json = (json for the next download instructions) -> chain3-zip_dir
+	// chain3-zip_dir - doesn't upgrade
+	require := s.Require()
+	home := copyTestData(s.T(), "download")
+	cfg := &cosmovisor.Config{Home: home, Name: "autod", AllowDownloadBinaries: true, PollInterval: 100, UnsafeSkipBackup: true, CustomPreupgrade: "missing.sh"}
+	logger := log.NewLogger(os.Stdout).With(log.ModuleKey, "cosmovisor")
+	upgradeFilename := cfg.UpgradeInfoFilePath()
+
+	// should run the genesis binary and produce expected output
+	currentBin, err := cfg.CurrentBin()
+	require.NoError(err)
+	require.Equal(cfg.GenesisBin(), currentBin)
+	launcher, err := cosmovisor.NewLauncher(logger, cfg)
+	require.NoError(err)
+
+	// Missing Preupgrade Script
+	stdout, stderr := newBuffer(), newBuffer()
+	args := []string{"some", "args", upgradeFilename}
+	_, err = launcher.Run(args, stdout, stderr)
+
+	require.Error(err)
+}
+
+// TestCustomPreupgrade will try running the script a few times and watch upgrades work properly
+// and args are passed through
+func (s *processTestSuite) TestLaunchProcessWithDownloadsAndPreupgrade() {
+	// test case upgrade path (binaries from testdata/download directory):
+	// genesis -> chain2-zip_bin
+	// chain2-zip_bin -> ref_to_chain3-zip_dir.json = (json for the next download instructions) -> chain3-zip_dir
+	// chain3-zip_dir - doesn't upgrade
+	require := s.Require()
+	home := copyTestData(s.T(), "download")
+	cfg := &cosmovisor.Config{Home: home, Name: "autod", AllowDownloadBinaries: true, PollInterval: 100, UnsafeSkipBackup: true, CustomPreupgrade: "preupgrade.sh"}
+	logger := log.NewLogger(os.Stdout).With(log.ModuleKey, "cosmovisor")
```

Comment:
Unless am mistaken but this logger is going to spam standard output. Could we please instead pass in a bytes.Buffer that we can even then later on check to ensure certain outputs are contained? If the results are immaterial, we can just use io.Discard